### PR TITLE
insights: add a moving daily snapshot for insights over all repos

### LINKED ISSUE #24186: Insights: add single frontend datapoint to backend insights so you don't have to wait a month to get "current" results
_This is a stretch issue, and if necessary the user can non-optimally solve this now by running the search in sourcegraph themself._

Because we now have a backend, if you wanted to "use the frontend" to help with insights, all we really need is a single present-time point. This means it doesn't matter how many repos you include because we can use the normal search API to just run the search. 

Visual concept of what I am imagining: 

image

This was never _not_ part of the "historical insights over all my repos" concept (there's no reason a user wouldn't want to see today but would want to see the first of the month) but it was cut for scope originally. 

I will defer to others as to whether we could get this done in time, but it would be awesome and our users would love it (and we need to do this in 3.32 if we don't do it in 3.31). The use case of "actively tracking a migration" means a user cares about these present-time points as well as the history. 

The considerations (reasons it might be trickier/not possible in the 2.5 days next week): 
- it needs to match filters that might be on the insight (so we can't just use the data series query; we also must honor the filters)
- it needs to match the backend (what this means, specifically, is that it should modify from just an "indexed search" to "an indexed search with `archived:yes fork:yes`attached, since that's what the backend currently does. Of course, if the insight data series query has a `fork:no` in it, this should override that).

### PR DESCRIPTION
**Pull request by coury-clark**
Closes #24186

Adds functionality to one of the code insights background processes to record a daily snapshot. This will allow us to keep the overall cardinality low with longer intervals, but still provide some amount of near real time visibility into the insight.

The daily snapshot will only be persisted for each series for one data point. Any previous snapshots for each series will be deleted in the same transaction as the new snapshot.

Introduces new `PersistMode` to the insight series store that will toggle a `recording` or `snapshot`.

### CODE REVIEW

File: enterprise/internal/insights/background/insight_enqueuer.go [hunk 1]
```diff
@@ -48,33 +50,54 @@ func newInsightEnqueuer(ctx context.Context, workerBaseStore *basestore.Store, i
...
+func enqueue(ctx context.Context, dataSeries []types.InsightSeries, mode store.PersistMode,
+	stampFunc func(ctx context.Context, insightSeries types.InsightSeries) (types.InsightSeries, error),
+	enqueueQueryRunnerJob func(ctx context.Context, job *queryrunner.Job) error) error {
+
 	// Deduplicate series that may be unique (e.g. different name/description) but do not have
 	// unique data (i.e. use the same exact search query or webhook URL.)
 	var (
 		uniqueSeries = map[string]types.InsightSeries{}
 		multi        error
-		offset       time.Duration
```
Comment (coury-clark) [inline enterprise/internal/insights/background/insight_enqueuer.go:77]: This processing is not rate limited separately, so removing all the old references to the worker `process_after`

File: enterprise/internal/insights/background/insight_enqueuer.go [hunk 2]
```diff
@@ -48,33 +50,54 @@ func newInsightEnqueuer(ctx context.Context, workerBaseStore *basestore.Store, i
...
+	if err != nil {
+		multi = multierror.Append(multi, err)
+	}
+
+	return multi
+}
+
+func enqueue(ctx context.Context, dataSeries []types.InsightSeries, mode store.PersistMode,
+	stampFunc func(ctx context.Context, insightSeries types.InsightSeries) (types.InsightSeries, error),
+	enqueueQueryRunnerJob func(ctx context.Context, job *queryrunner.Job) error) error {
```
Comment (slimsag) [inline enterprise/internal/insights/background/insight_enqueuer.go:94]: might make sense to change this to `func enqueue(ctx context.Context, job enqueueJob) error` and then define an `type enqueueJob struct` with these params to improve readability.

File: migrations/codeinsights/1000000012_insights_snapshots.up.sql [hunk 3]
```diff
@@ -0,0 +1,24 @@
+
+BEGIN;
+
+-- Insert migration here. See README.md. Highlights:
+--  * Always use IF EXISTS. eg: DROP TABLE IF EXISTS global_dep_private;
+--  * All migrations must be backward-compatible. Old versions of Sourcegraph
+--    need to be able to read/write post migration.
+--  * Historically we advised against transactions since we thought the
+--    migrate library handled it. However, it does not! /facepalm
```
Comment (slimsag) [inline migrations/codeinsights/1000000012_insights_snapshots.up.sql:9]: you can delete these comments FYI, they're only really useful when writing the migration.

**Review by slimsag — APPROVED** [review]
very cool!

**Comment by slimsag** [comment]
just out of curiosity: what made you choose going the route of an additional table instead of a field on the `series_points` table? just preference, or something else? (curious to learn for myself)

**Comment by coury-clark** [comment]
> just out of curiosity: what made you choose going the route of an additional table instead of a field on the `series_points` table? just preference, or something else? (curious to learn for myself)

To be honest, I think either way would be okay. At the end of the day it came down to the following for me:
1. A huge majority of the records are going to be type `record` instead of type `snapshot`, so storing data on all of those records is wasteful. We could get around that by using nullable columns, but I like the idea that we can strongly associate every record with useful metadata rather than having `null` be a pseudo-type.
2. When I think about how code insights will manifest on dot-com, I think the architecture will likely involve separate workstreams of background processing that coalesce through a higher-level view. This marginally inches towards that kind of architecture, where we may have a completely different inputs (could be something like Spark, Kafka, whatever) and we can do similar behaviors without having to support one schema for each environment. I could be wrong on this one, but I think this is a bet that will pay off in the end.

**Comment by coury-clark** [comment]
Failing check on the build is Percy

File: migrations/codeinsights/1000000012_insights_snapshots.up.sql (hunk 3 above)
Comment (coury-clark) [reply migrations/codeinsights/1000000012_insights_snapshots.up.sql:9]: Good point, not really sure why I'm leaving those.

We have to do some work on these in the near future to migrate away from Timescale, when we pass through these scripts we will clean them up then.

**Pull request merged**